// grunt-ts/tasks/modules/transformers.js
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var fs = require('fs');
var path = require('path');
var grunt = require('grunt');
var _str = require('underscore.string');
var _ = require('underscore');
var os = require('os');
var utils = require('./utils');

var currentTargetFiles;
var currentTargetDirs;

function getImports(currentFilePath, name, targetFiles, targetDirs, getIndexIfDir) {
    if (typeof getIndexIfDir === "undefined") { getIndexIfDir = true; }
    var files = [];

    var targetFile = _.find(targetFiles, function (targetFile) {
        return path.basename(targetFile) === name || path.basename(targetFile, '.d.ts') === name || path.basename(targetFile, '.ts') === name;
    });
    if (targetFile) {
        files.push(targetFile);
    }

    var targetDir = _.find(targetDirs, function (targetDir) {
        return path.basename(targetDir) === name;
    });
    if (targetDir) {
        var possibleIndexFilePath = path.join(targetDir, 'index.ts');

        if (getIndexIfDir && fs.existsSync(possibleIndexFilePath) && path.relative(currentFilePath, possibleIndexFilePath) !== '') {
            files.push(path.join(targetDir, 'index.ts'));
        } else {
            var filesInDir = utils.getFiles(targetDir, function (filename) {
                if (path.relative(currentFilePath, filename) === '') {
                    return true;
                }

                return path.extname(filename) && (!_str.endsWith(filename, '.ts') || _str.endsWith(filename, '.d.ts')) && !fs.lstatSync(filename).isDirectory();
            });
            filesInDir.sort();
            files = files.concat(filesInDir);
        }
    }

    return files;
}

function getTargetFolders(targetFiles) {
    var folders = [];
    _.forEach(targetFiles, function (targetFile) {
        var dir = path.dirname(targetFile);
        while (dir !== '.') {
            folders.push(dir);
            dir = path.dirname(dir);
        }
    });

    return folders;
}

var BaseTransformer = (function () {
    function BaseTransformer(key) {
        this.key = key;
        this.intro = BaseTransformer.tsSignature + key;
        this.match = new RegExp(utils.format(BaseTransformer.tsSignatureMatch, key));
        this.signatureGenerated = ' ' + this.intro + ':generated';
    }
    BaseTransformer.prototype.isGenerated = function (line) {
        return _str.contains(line, this.signatureGenerated);
    };

    BaseTransformer.prototype.isSignature = function (line) {
        return _str.contains(line, this.intro);
    };

    BaseTransformer.prototype.transform = function (sourceFile, config, outputLines) {
        throw new Error('Must override transform function');
    };
    BaseTransformer.tsSignature = '///ts:';
    BaseTransformer.tsSignatureMatch = '///ts:{0}=(.*)';
    return BaseTransformer;
})();

var ImportTransformer = (function (_super) {
    __extends(ImportTransformer, _super);
    function ImportTransformer() {
        _super.call(this, 'import');
        this.importError = '/// No glob matched name: ';
        this.template = _.template('import <%=filename%> = require(\'<%= pathToFile %>\');' + this.signatureGenerated);
    }
    ImportTransformer.prototype.transform = function (sourceFile, config, outputLines) {
        var _this = this;
        name = config;
        var sourceFileDirectory = path.dirname(sourceFile);

        var imports = getImports(sourceFile, name, currentTargetFiles, currentTargetDirs);

        if (imports.length) {
            _.forEach(imports, function (completePathToFile) {
                var filename = path.basename(completePathToFile, '.ts');

                if (filename.toLowerCase() === 'index') {
                    filename = path.basename(path.dirname(completePathToFile));
                }
                var pathToFile = './' + utils.makeRelativePath(sourceFileDirectory, completePathToFile.replace('.ts', ''));
                outputLines.push(_this.template({ filename: filename, pathToFile: pathToFile }));
            });
        } else {
            outputLines.push(this.importError + name + this.signatureGenerated);
        }
    };
    return ImportTransformer;
})(BaseTransformer);

var ExportTransformer = (function (_super) {
    __extends(ExportTransformer, _super);
    function ExportTransformer() {
        _super.call(this, 'export');
        this.importError = '/// No glob matched name: ';
        this.template = _.template('import <%=filename%>_file = require(\'<%= pathToFile %>\');' + this.signatureGenerated + os.EOL + 'export var <%=filename%> = <%=filename%>_file;' + this.signatureGenerated);
    }
    ExportTransformer.prototype.transform = function (sourceFile, config, outputLines) {
        var _this = this;
        name = config;
        var sourceFileDirectory = path.dirname(sourceFile);

        var imports = getImports(sourceFile, name, currentTargetFiles, currentTargetDirs, false);

        if (imports.length) {
            _.forEach(imports, function (completePathToFile) {
                var filename = path.basename(completePathToFile, '.ts');

                if (filename.toLowerCase() === 'index') {
                    filename = path.basename(path.dirname(completePathToFile));
                }
                var pathToFile = './' + utils.makeRelativePath(sourceFileDirectory, completePathToFile.replace('.ts', ''));
                outputLines.push(_this.template({ filename: filename, pathToFile: pathToFile }));
            });
        } else {
            outputLines.push(this.importError + name + this.signatureGenerated);
        }
    };
    return ExportTransformer;
})(BaseTransformer);

function transformFiles(changedFiles, targetFiles, target, task) {
    currentTargetDirs = getTargetFolders(targetFiles);
    currentTargetFiles = targetFiles;

    var transformers = [
        new ImportTransformer(),
        new ExportTransformer()
    ];

    _.forEach(changedFiles, function (fileToProcess) {
        var contents = fs.readFileSync(fileToProcess).toString();

        if (!_str.contains(contents, BaseTransformer.tsSignature)) {
            return;
        }

        var lines = contents.split(/\r\n|\r|\n/);
        var outputLines = [];

        for (var i = 0; i < lines.length; i++) {
            var line = lines[i];

            if (_.some(transformers, function (transformer) {
                return transformer.isGenerated(line);
            })) {
                continue;
            }

            if (_.some(transformers, function (transformer) {
                if (transformer.isSignature(line)) {
                    outputLines.push(line);

                    var match = line.match(transformer.match);
                    if (!match || !match[1]) {
                        outputLines.push('/// Must match: ' + transformer.match + transformer.signatureGenerated);
                    } else {
                        var config = match[1];
                        config = config.trim();

                        transformer.transform(fileToProcess, config, outputLines);
                    }

                    return true;
                }
            })) {
                continue;
            }

            outputLines.push(line);
        }
        var transformedContent = outputLines.join(os.EOL);
        if (transformedContent !== contents) {
            grunt.file.write(fileToProcess, transformedContent);
        }
    });
}
exports.transformFiles = transformFiles;
